Adding Switchable Boundary Conditions

tanoret opened the pull request:
Closes #25803

tanoret committed:
adding switch bcs Refs #25803

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1834345/) on dea91d4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25806/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1834385/) on dea91d4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3d2af0</th>
      <th colspan=4>#25806 dea91d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.63%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92355</td>
      <td>92354</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15498</td>
      <td>15499</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25806/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25806/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>3d2af0</th>
      <th colspan=4>#25806 dea91d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.59%</td>
      <td>84.21%</td>
      <td>-0.38%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12315</td>
      <td>12315</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2244</td>
      <td>2309</td>
      <td>+65</td>
      <td>65</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25806/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25806/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25806/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad reviewed (changes requested):
It looks like you're missing a `tests` file

lindsayad reviewed (changes requested):
my question is: can't we already do this without new objects by doing boolean control of the `MooseObject` `enable` parameter?

GiudGiud commented:
I think the problem with that would be that the dirichlet BCs are cached by the variable (or the fi? cant remember), and I dont know that we check whether they are active before caching them or how often we cache

tanoret committed:
adding test file Refs #25806

tanoret commented:
> I think the problem with that would be that the dirichlet BCs are cached by the variable (or the fi? cant remember), and I dont know that we check whether they are active before caching them or how often we cache

This is working fine for the tests I need to do for domain overlapping. The nl boundary residual should be reconstructed at per time_step, which is what we need for domain overlap 
![flow_inversion](https://github.com/idaholab/moose/assets/29891082/740311a6-08c1-467a-93eb-5706a99fd204)

lindsayad commented:
> I think the problem with that would be that the dirichlet BCs are cached by the variable (or the fi? cant remember), and I dont know that we check whether they are active before caching them or how often we cache

We worked on this in https://github.com/idaholab/moose/pull/24446 for this kind of problem. We changed the caching of Dirichlet conditions to occur on timestep setup to accomadate changes in the `enable` status. Look at the test input in that PR; it has some commonality with the test input in this PR

lindsayad commented:
> This is working fine for the tests I need to do for domain overlapping

I believe that it works. I just want to understand if what we already have doesn't work in order to potentially avoid repeating ourselves

tanoret commented:
> I believe that it works. I just want to understand if what we already have doesn't work in order to potentially avoid repeating ourselves

Trying this now. Waiting for the code to recompile. I kind of like the `switch=on/off` because it is easy to understand in the action control logic. Also, if the BC is `off` I can decide what value I assign to the phase to potentially add limiters and correct for mass imbalance (not doing this yet)

tanoret commented:
> Trying this now. Waiting for the code to recompile. I kind of like the `switch=on/off` because it is easy to understand in the action control logic. Also, if the BC is `off` I can decide what value I assign to the phase to potentially add limiters and correct for mass imbalance (not doing this yet)

It is not working for me. Here is my input in case that I am missing anything. I think that the issue is that the contribution to the Jacobian structure is fixed in the current method. So, we can switch boundary values but not really boundary types (I may be wrong on this, haven't go much in depth)

`rho = 'rho'
l = 10
inlet_area = 1
velocity_interp_method = 'rc'
advected_interp_method = 'average'

# Artificial fluid properties
# For a real case, use a GeneralFluidFunctorProperties and a viscosity rampdown
# or initialize very well!
k = 1
cp = 1000
mu = 1e2

# Operating conditions
inlet_temp = 300
outlet_pressure = 1e5
inlet_velocity = 0.001

end_time = 2.0
switch_time = 1.0

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = ${l}
    ymin = 0
    ymax = 1
    nx = 10
    ny = 5
  []
[]

[GlobalParams]
  rhie_chow_user_object = 'rc'
[]

[UserObjects]
  [rc]
    type = INSFVRhieChowInterpolator
    u = u
    v = v
    pressure = pressure
  []
[]

[Variables]
  [u]
    type = INSFVVelocityVariable
    initial_condition = ${inlet_velocity}
  []
  [v]
    type = INSFVVelocityVariable
    initial_condition = 1e-15
  []
  [pressure]
    type = INSFVPressureVariable
    initial_condition = ${outlet_pressure}
  []
  [T]
    type = INSFVEnergyVariable
    initial_condition = ${inlet_temp}
  []
[]

[AuxVariables]
  [power_density]
    type = MooseVariableFVReal
    initial_condition = 1e4
  []
[]

[FVKernels]
  [mass_time]
    type = WCNSFVMassTimeDerivative
    variable = pressure
    drho_dt = drho_dt
  []
  [mass]
    type = INSFVMassAdvection
    variable = pressure
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    rho = ${rho}
  []

  [u_time]
    type = WCNSFVMomentumTimeDerivative
    variable = u
    drho_dt = drho_dt
    rho = rho
    momentum_component = 'x'
  []
  [u_advection]
    type = INSFVMomentumAdvection
    variable = u
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
    rho = ${rho}
    momentum_component = 'x'
  []
  [u_viscosity]
    type = INSFVMomentumDiffusion
    variable = u
    mu = ${mu}
    momentum_component = 'x'
  []
  [u_pressure]
    type = INSFVMomentumPressure
    variable = u
    momentum_component = 'x'
    pressure = pressure
  []

  [v_time]
    type = WCNSFVMomentumTimeDerivative
    variable = v
    drho_dt = drho_dt
    rho = rho
    momentum_component = 'y'
  []
  [v_advection]
    type = INSFVMomentumAdvection
    variable = v
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
    rho = ${rho}
    momentum_component = 'y'
  []
  [v_viscosity]
    type = INSFVMomentumDiffusion
    variable = v
    mu = ${mu}
    momentum_component = 'y'
  []
  [v_pressure]
    type = INSFVMomentumPressure
    variable = v
    momentum_component = 'y'
    pressure = pressure
  []

  [temp_time]
    type = WCNSFVEnergyTimeDerivative
    variable = T
    rho = rho
    drho_dt = drho_dt
  []
  [temp_conduction]
    type = FVDiffusion
    coeff = 'k'
    variable = T
  []
  [temp_advection]
    type = INSFVEnergyAdvection
    variable = T
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
  []
  [heat_source]
    type = FVCoupledForce
    variable = T
    v = power_density
  []
[]

[FVBCs]
  # Inlet
  [inlet_u]
    type = WCNSFVSwitchableInletVelocityBC
    variable = u
    boundary = 'left'
    mdot_pp = 'inlet_mdot'
    area_pp = 'surface_inlet'
    rho = 'rho'
    switch = true
  []
  [outlet_u]
    type = WCNSFVSwitchableInletVelocityBC
    variable = u
    boundary = 'right'
    mdot_pp = 'inlet_mdot'
    area_pp = 'surface_inlet'
    rho = 'rho'
    switch = false
    scaling_factor = -1.0
  []

  [inlet_v]
    type = WCNSFVInletVelocityBC
    variable = v
    boundary = 'left'
    mdot_pp = 0
    area_pp = 'surface_inlet'
    rho = 'rho'
  []

  [inlet_T]
    type = WCNSFVInletTemperatureBC
    variable = T
    boundary = 'left'
    temperature_pp = 'inlet_T'
  []
  [outlet_T]
    type = NSFVOutflowTemperatureBC
    variable = T
    boundary = 'right'
    u = u
    v = v
    rho = 'rho'
    cp = 'cp'
    backflow_T = ${inlet_temp}
  []

  [outlet_p]
    type = INSFVSwitchableOutletPressureBC
    variable = pressure
    boundary = 'right'
    function = ${outlet_pressure}
    switch = true
  []
  [inlet_p]
    type = INSFVSwitchableOutletPressureBC
    variable = pressure
    boundary = 'left'
    function = ${outlet_pressure}
    switch = false
  []

  # Walls
  [no_slip_x]
    type = INSFVNoSlipWallBC
    variable = u
    boundary = 'top bottom'
    function = 0
  []
  [no_slip_y]
    type = INSFVNoSlipWallBC
    variable = v
    boundary = 'top bottom'
    function = 0
  []
[]

[Functions]
  [conditional_function]
    type = ParsedFunction
    expression = 't > ${switch_time}'
  []
[]

[Controls]
  [p_threshold]
    type = ConditionalFunctionEnableControl
    conditional_function = conditional_function
    disable_objects = 'FVBCs::inlet_u FVBCs::outlet_p'
    enable_objects = 'FVBCs::outlet_u FVBCs::inlet_p'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]

# used for the boundary conditions in this example
[Postprocessors]
  [inlet_mdot]
    type = Receiver
    default = '${fparse 1980 * inlet_velocity * inlet_area}'
  []
  [surface_inlet]
    type = AreaPostprocessor
    boundary = 'left'
    execute_on = 'INITIAL'
  []
  [inlet_T]
    type = Receiver
    default = ${inlet_temp}
  []
  [outlet_mfr]
    type = VolumetricFlowRate
    boundary = 'right'
    advected_quantity = 1.0
    vel_x = u
    vel_y = v
  []
[]

[FluidProperties]
  [fp]
    type = FlibeFluidProperties
  []
[]

[Materials]
  [const_functor]
    type = ADGenericFunctorMaterial
    prop_names = 'cp k'
    prop_values = '${cp} ${k}'
  []
  [rho]
    type = RhoFromPTFunctorMaterial
    fp = fp
    temperature = T
    pressure = pressure
  []
  [ins_fv]
    type = INSFVEnthalpyMaterial
    temperature = 'T'
    rho = ${rho}
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -pc_factor_shift_type'
  petsc_options_value = 'lu       NONZERO'
  dt = 0.1
  end_time = ${end_time}

  nl_abs_tol = 1e-12
  nl_max_its = 50
  line_search = 'none'

  automatic_scaling = true
[]

[Outputs]
  csv = true
  execute_on = 'TIMESTEP_END'
[]
`

lindsayad commented:
See https://github.com/idaholab/moose/pull/25806/commits/e7ae171a3a8f1a9a5788cfa7563192e7daf3f581; it uses existing objects and matches the gold file you added

lindsayad commented:
If you like the `switch` parameter name better than `enable`, then we can actually change it in the `navier_stokes` objects, using the `InputParameters::renameParam` API. I personally think the `enable` parameter more clearly conveys the intent here as when I read `switch = true` I read that we should be flipping a switch; if it was something like `switch = on`, then that would be clear. But I'm open to discussion

tanoret commented:
> If you like the `switch` parameter name better than `enable`, then we can actually change it in the `navier_stokes` objects, using the `InputParameters::renameParam` API. I personally think the `enable` parameter more clearly conveys the intent here as when I read `switch = true` I read that we should be flipping a switch; if it was something like `switch = on`, then that would be clear. But I'm open to discussion

Yes, agree with this. I can do an enumerator for `switch = on/off`, no problem. I do prefer to have my switchable boundary conditions for two reasons:
1. I can control what is happening if the boundary condition is `off`, i.e., one can understand directly what the code is doing and eventually add limiters or other needed functionalities for coupling.
2. Somehow related, I think it is clearer for the future user to understand what is going on when reading the switchable boundary conditions without going many levels of inheritance below the required on/off functionality

What do you think?

tanoret commented:
Thinking again about this, we don't have a nice way to switch on/off the switcher. I will fold back to `enable` for simplicity. I just wished that `enable` appears in the source code of the switchable boundary conditions for people to understand better the code when they read it...

tanoret committed:
changing switch -> enable Refs #25806

tanoret commented:
Changed switch --> enable and added a comment in the .C files so that the users know what is happening. It could be cleaner and more straightforward to read if the `switch` parameter was directly declared in the switchable bcs source code, but I agree that this will duplicate the `enable` parameter

lindsayad commented:
> I just wished that `enable` appears in the source code of the switchable boundary conditions for people to understand better the code when they read it...

That's fair. It can definitely be some work to figure out what all is available in a derived class. I think the doxygen pages are pretty helpful for data members and the moose docs page is pretty helpful for input parameters, but nothing would beat being able to open the class files and see everything there in terms of readability.

lindsayad commented:
> I do prefer to have my switchable boundary conditions for two reasons:

Ok I can get on board with the new classes, but I would vote for the following change: the `WCNSFVSwitchableInletVelocityBC` class should inherit directly from `WCNSFVInletVelocityBC` and similarly `INSFVSwitchableOutletPressureBC` inherit directly from `INSFVOutletPressureBC`. Now this goes against "readability of a single class" but it helps us maximize code re-use. In such a way if someone improves or fixes a bug in `WCNSFVInletVelociytBC/INSFVOutletPressureBC` then the switchable classes immediately benefit.

With this inheritance structure then I would have the `boundaryValue` methods look like this:

### velocity

```c++
ADReal
WCNSFVSwitchableInletVelocityBC::boundaryValue(const FaceInfo & fi) const
{
  /// Enable is a parameter coming from MooseObject                                                                   
  /// It allows to enable/disable an object                                                                           
  /// The user can think of enable as an on/off switch for the boundary condition                                     
  /// In this case, if enable is true, we activate the boundary condition                                             
  /// Otherwise, we perform a restricted face internal interpolation                                                  
  if (_enabled)
    return WCNSFVInletVelocityBC::boundaryValue(fi);
  else
    return _var.getExtrapolatedBoundaryFaceValue(fi, false, false, fi.elemPtr(), determineState()) *
           _face_limiter;
}
```

### pressure

```c++
ADReal
INSFVSwitchableOutletPressureBC::boundaryValue(const FaceInfo & fi) const
{
  /// Enable is a parameter coming from MooseObject                                                                   
  /// It allows to enable/disable an object                                                                           
  /// The user can think of enable as an on/off switch for the boundary condition                                     
  /// In this case, if enable is true, we activate the boundary condition                                             
  /// Otherwise, we perform a restricted face internal interpolation                                                  
  if (_enabled)
    return INSFVOutletPressureBC::boundaryValue(fi);
  else
    return _var.getExtrapolatedBoundaryFaceValue(fi, false, false, fi.elemPtr(), determineState()) *
           _face_limiter;
}
```

loganharbour commented:
Just popping in after seeing the super long e-mail chain - isn't `_enabled` something that's used by a system instead of in an object?

lindsayad commented:
Hmm yea boundaryValue should actually never be called if _enabled is false. So @tanoret we will need our own boolean parameter like the ‘switch’ you had originally. I agree that with that parameter name an enum with on and off would be nicer but that requires development unlike a boolean controller, so I’m fine with a boolean

tanoret committed:
enable->switch + modified inheritance Refs #25806

tanoret commented:
> Hmm yea boundaryValue should actually never be called if _enabled is false. So @tanoret we will need our own boolean parameter like the ‘switch’ you had originally. I agree that with that parameter name an enum with on and off would be nicer but that requires development unlike a boolean controller, so I’m fine with a boolean

Thanks. Changed back to bool and corrected the inheritance of the classes.

lindsayad commented on modules/navier_stokes/doc/content/source/fvbcs/INSFVSwitchableOutletPressureBC.md:
```diff
@@ -0,0 +1,32 @@
+
+The file is similar to [INSFVOutletPressureBC.md] but it allows the boundary conditions to be switched on/off.
+
+This object simply wraps [`FVFunctionDirichletBC`](FVFunctionDirichletBC.md), so
```
Need to update this documentation

lindsayad replied:
Actually I see this is a copy of the now-base-class doc, so you can just delete this whole paragraph.

tanoret replied:
Thanks. Deleted

lindsayad commented on modules/navier_stokes/doc/content/source/fvbcs/WCNSFVSwitchableInletVelocityBC.md:
```diff
@@ -0,0 +1,53 @@
+
+The file is similar to [WCNSFVInletVelocityBC.md] but it allows the boundary conditions to be switched on/off.
+
+There are two options for specifying a component of the inlet velocity:
+
+- specifying a velocity postprocessor
+
+- specifying a mass flow rate postprocessor and a density functor. The functor is
+  usually a functor material property, defined by a [GeneralFunctorFluidProps.md].
+  The scaling factor can be used to account for projections if the inlet flow and
+  the surface are not aligned.
+
+
+This boundary condition works with postprocessors, which may be replaced by constant
+values in the input. The intended use case for this boundary condition is to be receiving its value from
+a coupled application, using a [Receiver.md] postprocessor.
```
Now that you are directly inheriting from `WCNSFVInletVelocityBC`, let's remove this given it exists in the base class doc

tanoret replied:
Thanks. Deleted

lindsayad commented on modules/navier_stokes/doc/content/source/fvbcs/WCNSFVSwitchableInletVelocityBC.md:
```diff
@@ -0,0 +1,53 @@
+
+The switch works as follows:
+
+- If `switch = true`: the boundary condition is applied as described above.
```
```suggestion
- If `switch = true`: the boundary condition is applied as described in [WCNSFVInletVelocityBC.md].
```

tanoret replied:
Thanks. Changed

lindsayad commented on modules/navier_stokes/doc/content/source/fvbcs/INSFVSwitchableOutletPressureBC.md:
```diff
@@ -0,0 +1,32 @@
+
+The switch works as follows:
+
+- If `switch = true`: the boundary condition is applied as described above.
```
```suggestion
- If `switch = true`: the boundary condition is applied as described in [INSFVOutletPressureBC.md].
```

tanoret replied:
Thanks. Changed

lindsayad commented on modules/navier_stokes/include/fvbcs/INSFVSwitchableOutletPressureBC.h:
```diff
@@ -0,0 +1,32 @@
+#include "INSFVOutletPressureBC.h"
+
+/**
+ * A class for setting the value of the pressure at an outlet of the system.
+ * It may not be used with a mean/pinned-pressure approach
```
This is just a copy of the base class doxy I assume. I don't know if you want to add anything about its "switchability"

lindsayad commented on modules/navier_stokes/include/fvbcs/INSFVSwitchableOutletPressureBC.h:
```diff
@@ -0,0 +1,32 @@
+  ADReal boundaryValue(const FaceInfo & /* fi */) const override;
+
+private:
+  /// Boolean switch to turn boudnary condition on/off
```
```suggestion
  /// Boolean switch to turn boundary condition on/off
```

tanoret replied:
Thanks. Corrected

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVSwitchableInletVelocityBC.h:
```diff
@@ -0,0 +1,32 @@
+#include "WCNSFVInletVelocityBC.h"
+
+/**
+ * Dirichlet boundary conditions for the velocity, set from either a velocity postprocessor
+ * or a mass flow rate divided by density and surface
```
This is just a copy of the base class doxy I assume. I don't know if you want to add anything about its "switchability"

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVSwitchableInletVelocityBC.h:
```diff
@@ -0,0 +1,32 @@
+protected:
+  ADReal boundaryValue(const FaceInfo & fi) const override;
+
+  /// Boolean switch to turn boudnary condition on/off
```
```suggestion
  /// Boolean switch to turn boundary condition on/off
```

tanoret replied:
Thanks. Corrected

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVSwitchableOutletPressureBC.C:
```diff
@@ -0,0 +1,57 @@
+  params.addParam<bool>("switch", true, "Switch on (true) / off (false) for boundary condition.");
+  params.declareControllable("switch");
+
+  params.addParam<Real>("face_limiter", 1.0, "Interpolation face flux limiter.");
```
Is "interpolation" an accurate descriptor given the boundary nature of this?

tanoret replied:
Thanks. Removed 'Interpolation'

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVSwitchableOutletPressureBC.C:
```diff
@@ -0,0 +1,57 @@
+    _switch_bc(getParam<bool>("switch")),
+    _face_limiter(getParam<Real>("face_limiter"))
+{
+  if (!dynamic_cast<INSFVPressureVariable *>(&_var))
+    paramError("variable",
+               "The variable argument to INSFVSwitchableOutletPressureBC must be of type "
+               "INSFVPressureVariable");
+
+  // Check parameters
+  if ((_functor && (_pp_value || _function)) || (_function && _pp_value) ||
+      (!_functor && !_pp_value && !_function))
+    mooseError("One and only one of function/functor/postprocessor may be specified for the outlet "
+               "pressure");
```
```suggestion
```
Already done in the base class

tanoret replied:
Thanks. Removed

lindsayad commented on modules/navier_stokes/src/fvbcs/WCNSFVSwitchableInletVelocityBC.C:
```diff
@@ -0,0 +1,60 @@
+  if (_switch_bc)
+    return WCNSFVInletVelocityBC::boundaryValue(fi) * _face_limiter;
+  else
+    return _var.getExtrapolatedBoundaryFaceValue(fi, false, false, fi.elemPtr(), determineState()) *
```
```suggestion
    return _var(singleSidedFaceArg(), determineState()) *
```
It is theoretically possible (although maybe not practically probable) that we have an outflow on an internal boundary of the mesh in which case the `fi.neighborPtr()` would be the cell we would want to extrapolate from

tanoret replied:
Thanks. Changed

tanoret replied:
Sorry. Come back on this one. I am getting a seg fault when trying to run with singleSidedFaceArg(). I am keeping as it is for now.

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVSwitchableOutletPressureBC.C:
```diff
@@ -0,0 +1,57 @@
+  if (_switch_bc)
+    return INSFVOutletPressureBC::boundaryValue(fi) * _face_limiter;
+  else
+    return _var.getExtrapolatedBoundaryFaceValue(fi, false, false, fi.elemPtr(), determineState()) *
```
```suggestion
    return _var(singleSidedFaceArg(), determineState()) *
```

lindsayad replied:
If you're worried this conveys less meaning to a reader (it almost certainly does), then feel free to comment that this will perform a boundary value extrapolation

tanoret replied:
Thanks. Changed to the functor interface. I think it is pretty clear your way

tanoret replied:
Sorry. Come back on this one. I am getting a seg fault when trying to run with singleSidedFaceArg(). I am keeping as it is for now.

lindsayad commented on framework/include/variables/MooseVariableFV.h:
If you follow my suggestions to use the functor interfaces instead of calling this variable method directly, then we can revert the changes in this file

lindsayad replied:
perhaps you meant to say you reverted this change. 👍

lindsayad commented on modules/navier_stokes/src/fvbcs/WCNSFVSwitchableInletVelocityBC.C:
```diff
@@ -0,0 +1,60 @@
+    _switch_bc(getParam<bool>("switch")),
+    _face_limiter(getParam<Real>("face_limiter"))
+{
+  if (!dynamic_cast<INSFVVelocityVariable *>(&_var))
+    paramError("variable",
+               "The variable argument to WCNSFVSwitchableInletVelocityBC must be of type "
+               "INSFVVelocityVariable");
+
+  // Density is often set as global parameters so it is not checked
+  if (_velocity_pp && (_mdot_pp || _area_pp))
+    mooseWarning("If setting the velocity directly, no need for inlet mass flow rate or area");
+
+  // Need enough information if trying to use a mass flow rate postprocessor
+  if (!_velocity_pp && (!_mdot_pp || !_area_pp || !_rho))
+    mooseError("Mass flow rate, area and density should be provided if velocity is not");
```
```suggestion
```

tanoret replied:
Thanks. Folded back to original.

tanoret replied:
Sorry. Changed this back to the original PR as oneSidedFaceArg() is seg-faulting for me

lindsayad replied:
this change should be unrelated to that

lindsayad replied:
looks like you didn't revert this change 👍

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+  []
+  [v]
+    type = INSFVVelocityVariable
+    initial_condition = 1e-15
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+    mdot_pp = 'inlet_mdot'
+    area_pp = 'surface_inlet'
+    rho = 'rho'
+    enable = true
```
```suggestion
    switch = true
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+    mdot_pp = 'inlet_mdot'
+    area_pp = 'surface_inlet'
+    rho = 'rho'
+    enable = false
```
```suggestion
    switch = false
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+    variable = pressure
+    boundary = 'right'
+    function = ${outlet_pressure}
+    enable = true
```
```suggestion
    switch = true
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+    variable = pressure
+    boundary = 'left'
+    function = ${outlet_pressure}
+    enable = false
```
```suggestion
    switch = false
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+[Controls]
+  [func_control_u_inlet]
+    type = BoolFunctionControl
+    parameter = 'FVBCs/inlet_u/enable'
```
```suggestion
    parameter = 'FVBCs/inlet_u/switch'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+  []
+  [func_control_u_outlet]
+    type = BoolFunctionControl
+    parameter = 'FVBCs/outlet_u/enable'
```
```suggestion
    parameter = 'FVBCs/outlet_u/switch'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+  []
+  [func_control_p_outlet]
+    type = BoolFunctionControl
+    parameter = 'FVBCs/outlet_p/enable'
```
```suggestion
    parameter = 'FVBCs/outlet_p/switch'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,383 @@
+  []
+  [func_control_p_inlet]
+    type = BoolFunctionControl
+    parameter = 'FVBCs/inlet_p/enable'
```
```suggestion
    parameter = 'FVBCs/inlet_p/switch'
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/tests:
```diff
@@ -1,10 +1,16 @@
+    type = CSVDiff
+    input = switch_vel_pres_bc.i
+    csvdiff = switch_vel_pres_bc_out.csv
+    requirement = 'The system shall be able to switch boundaries during runtime.'
```
```suggestion
    requirement = 'The system shall be able to switch the behavior of velocity and pressure boundary conditions during runtime.'
```

lindsayad reviewed (changes requested):
Thanks for making the inheritance changes!

tanoret committed:
addressing review comments Refs #25806

lindsayad committed:
Avoid fully developed flow bc error

lindsayad committed:
Comments on infinite recursion

And relax the assertion in getExtrapolatedBoundaryFaceValue

lindsayad commented:
@tanoret I've pushed a couple commits that fix the debug failure. I hope you don't mind

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1843022/) on f450040 : invalidated by @lindsayad

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1843012/) on f450040 : invalidated by @lindsayad

lindsayad commented:
@GiudGiud could you review my two commits please?

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/controls/switch-pressure-bc/switch_vel_pres_bc.i:
```diff
@@ -0,0 +1,382 @@
+[Controls]
+  [func_control_u_inlet]
+    type = BoolFunctionControl
+    parameter = 'FVBCs/inlet_u/switch'
```
if you called it something more definite than just switch you could rely on this behavior for bool function control:
```
  params.addRequiredParam<std::string>(
      "parameter",
      "The input parameter(s) to control. Specify a single parameter name and all "
      "parameters in all objects matching the name will be updated");
```

GiudGiud replied:
actually you would need to use a 'flip' behavior, because they have the same name but you are using different values

tanoret replied:
Thanks. Changes `switch` --> `switch_bc`

GiudGiud commented on modules/navier_stokes/doc/content/source/fvbcs/INSFVSwitchableOutletPressureBC.md:
```diff
@@ -0,0 +1,22 @@
+- If `switch = true`: the boundary condition is applied as described in [INSFVOutletPressureBC.md].
+
+- If `switch = false`: the boundary condition is not applied and a single sided extrapolation to the boundary
+  is applied from internal extrapolation. The user can expect second order convergence.
```
can the user expect second order convergence since we use single-term boundary expansion?

tanoret replied:
Agree. Removed the expectation for second order convergence in the text

tanoret committed:
changing switch flag Refs #25806
